### PR TITLE
feat: implementing pip-7

### PR DESCRIPTION
**Pull request by amirvalhalla**
## Description

This PR implemented [pip-7](https://pips.pactus.org/PIPs/pip-7)

### CODE REVIEW

**Review by kehiy — APPROVED** [review]
@amirvalhalla Thanks, everything same as pip-7.

File: sync/bundle/message/hello_test.go [hunk 1]
```diff
@@ -15,6 +16,17 @@ func TestHelloType(t *testing.T) {
 	assert.Equal(t, m.Type(), TypeHello)
 }
 
+func TestHelloMessageMyTimeUnix(t *testing.T) {
+	myTime := time.Now()
```
Comment (b00f) [inline sync/bundle/message/hello_test.go:20]: We can write a better test;
create time1 before NewHelloMessage(...)

Then time in HelloMessage should be GreaterOrEqual than `time1` and SmalerOrEqual Now.

**Review by b00f — CHANGES_REQUESTED** [review]

File: sync/handler_hello.go [hunk 2]
```diff
@@ -38,6 +39,14 @@ func (handler *helloHandler) ParseMessage(m message.Message, initiator peer.ID)
 		return handler.acknowledge(response, initiator)
 	}
 
+	if time.Since(msg.MyTime()) > 10*time.Second {
+		response := message.NewHelloAckMessage(message.ResponseCodeRejected,
+			fmt.Sprintf("hello message difference time is more than 10 seconds,"+
```
Comment (b00f) [inline sync/handler_hello.go:44]: Messages might be slow to reach the node, so let's just say the times on the two nodes are different by more than 10 seconds: `time discrepancy exceeds 10 seconds`.

File: sync/handler_hello_test.go [hunk 3]
```diff
@@ -44,8 +45,24 @@ func TestParsingHelloMessages(t *testing.T) {
 
 			assert.NoError(t, td.receivingNewMessage(td.sync, msg, pid))
 			td.checkPeerStatus(t, pid, peerset.StatusCodeBanned)
-			bundle := td.shouldPublishMessageWithThisType(t, td.network, message.TypeHelloAck)
-			assert.Equal(t, bundle.Message.(*message.HelloAckMessage).ResponseCode, message.ResponseCodeRejected)
+			bdl := td.shouldPublishMessageWithThisType(t, td.network, message.TypeHelloAck)
+			assert.Equal(t, bdl.Message.(*message.HelloAckMessage).ResponseCode, message.ResponseCodeRejected)
+		})
+
+	t.Run("Receiving Hello message from a peer. Difference of Hello Message Time with node time is a lot.",
```
Comment (b00f) [inline sync/handler_hello_test.go:52]: We need two tests:
In the first test, `my_time` is: `Now() - 10 seconds`.
In the second test, `my_time` is: `Now() + 20 seconds`.

I purposely added an extra 10 seconds to the time to support running tests in slow CI/CD actions.

**Review by b00f — CHANGES_REQUESTED** [review]

File: sync/bundle/message/hello_test.go [hunk 4]
```diff
@@ -45,6 +46,26 @@ func TestHelloMessage(t *testing.T) {
...
+		time1 := time.Now()
+		myTimeUnixMilli := time1.UnixMilli()
+
+		m := NewHelloMessage(ts.RandPeerID(), "Alice", 100, 0, ts.RandHash(), ts.RandHash())
+
+		assert.LessOrEqual(t, m.MyTimeUnixMilli, time.Now().UnixMilli())
+		assert.GreaterOrEqual(t, m.MyTimeUnixMilli, myTimeUnixMilli)
+	})
+
+	t.Run("MyTimeUnixMilli of time1 is greater than hello message time", func(t *testing.T) {
```
Comment (b00f) [inline sync/bundle/message/hello_test.go:59]: This test is redundant. right? We almost cover every thing in above test.

**Review by b00f — APPROVED** [review]

**Comment by b00f** [comment]
@amirvalhalla Looks good. Just one tests looks extra.

File: sync/bundle/message/hello_test.go (hunk 4 above)
Comment (amirvalhalla) [reply sync/bundle/message/hello_test.go:59]: @b00f 
yes, I've forgot to remove this test, I'll fix it right now

**Review by b00f — APPROVED** [review]

**Review by b00f — APPROVED** [review]

**Pull request merged**